qt/Makefile

```diff
@@ -1,4 +1,12 @@
-SHELL := bash
+SHELL := /bin/bash
+FIND := $(if $(wildcard /bin/find),/bin/find,/usr/bin/find)
+
+ifeq ($(OS),Windows_NT)
+	INSTALL_PYAUDIO := python -m pip install git+https://github.com/evandroforks/pyaudio
+else
+	INSTALL_PYAUDIO := @echo Skipping pyaudio for Windows...
+endif
```

Comment:
Please remove this, I feel more comfortable linking users to the wheels that Anki is built with.

Reply:
If I remove this `python -m pip install git+https...` Python tries to install it from pypi, and the installation breaks with this error:
```
make[1]: Leaving directory '/cygdrive/f/anki/pylib'
make[1]: Entering directory '/cygdrive/f/anki/qt'
python -m pip install -e .
Obtaining file:///F:/anki/qt
Requirement already satisfied: beautifulsoup4 in f:\anki\pyenv\lib\site-packages (from aqt==2.1.21) (4.8.2)
Requirement already satisfied: requests in f:\anki\pyenv\lib\site-packages (from aqt==2.1.21) (2.23.0)
Collecting send2trash
  Using cached Send2Trash-1.5.0-py3-none-any.whl (12 kB)
Collecting pyaudio
  Using cached PyAudio-0.2.11.tar.gz (37 kB)
Collecting markdown
  Using cached Markdown-3.2.1-py2.py3-none-any.whl (88 kB)
Collecting jsonschema
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Requirement already satisfied: psutil in f:\anki\pyenv\lib\site-packages (from aqt==2.1.21) (5.7.0)
Collecting pywin32
  Using cached pywin32-227-cp38-cp38-win_amd64.whl (9.1 MB)
Requirement already satisfied: soupsieve>=1.2 in f:\anki\pyenv\lib\site-packages (from beautifulsoup4->aqt==2.1.21) (2.0)
Requirement already satisfied: certifi>=2017.4.17 in f:\anki\pyenv\lib\site-packages (from requests->aqt==2.1.21) (2019.11.28)
Requirement already satisfied: idna<3,>=2.5 in f:\anki\pyenv\lib\site-packages (from requests->aqt==2.1.21) (2.9)
Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in f:\anki\pyenv\lib\site-packages (from requests->aqt==2.1.21) (1.25.8)
Requirement already satisfied: chardet<4,>=3.0.2 in f:\anki\pyenv\lib\site-packages (from requests->aqt==2.1.21) (3.0.4)
Requirement already satisfied: setuptools>=36 in f:\anki\pyenv\lib\site-packages (from markdown->aqt==2.1.21) (45.2.0)
Processing c:\users\professional\appdata\local\pip\cache\wheels\be\48\69\322d0d62b8a2eb62b4b497b10c30902d0d2cb4599ca0d4e1ba\pyrsistent-0.15.7-cp38-cp38-win_amd64.whl
Requirement already satisfied: attrs>=17.4.0 in f:\anki\pyenv\lib\site-packages (from jsonschema->aqt==2.1.21) (19.3.0)
Requirement already satisfied: six>=1.11.0 in f:\anki\pyenv\lib\site-packages (from jsonschema->aqt==2.1.21) (1.14.0)
Building wheels for collected packages: pyaudio
  Building wheel for pyaudio (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'F:\anki\pyenv\Scripts\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Professional\\AppData\\Local\\Temp\\pip-install-_np3kykb\\pyaudio\\setup.py'"'"'; __file__='"'"'C:\\Users\\Professional\\AppData\\Local\\Temp\\pip-install-_np3kykb\\pyaudio\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\Professional\AppData\Local\Temp\pip-wheel-pl7l_cs_'
       cwd: C:\Users\Professional\AppData\Local\Temp\pip-install-_np3kykb\pyaudio\
  Complete output (53 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.8
  copying src\pyaudio.py -> build\lib.win-amd64-3.8
  running build_ext
  building '_portaudio' extension
  creating build\temp.win-amd64-3.8
  creating build\temp.win-amd64-3.8\Release
  creating build\temp.win-amd64-3.8\Release\src
  F:\VisualStudio2017\VC\Tools\MSVC\14.12.25827\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DMS_WIN64=1 -IF:\anki\pyenv\include -IF:\Python\include -IF:\Python\include -IF:\VisualStudio2017\VC\Tools\MSVC\14.12.25827\ATLMFC\include -IF:\VisualStudio2017\VC\Tools\MSVC\14.12.25827\include "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" /Tcsrc/_portaudiomodule.c /Fobuild\temp.win-amd64-3.8\Release\src/_portaudiomodule.obj
  _portaudiomodule.c
  f:\python\include\pyconfig.h(117): warning C4005: 'MS_WIN64': macro redefinition
  f:\python\include\pyconfig.h(117): note: command-line arguments:  see previous definition of 'MS_WIN64'
  src/_portaudiomodule.c(44): warning C4005: 'min': macro redefinition
  C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt\stdlib.h(1290): note: see previous definition of 'min'
  src/_portaudiomodule.c(1367): error C2059: syntax error: '{'
  src/_portaudiomodule.c(1367): warning C4013: '__typeof__' undefined; assuming extern returning int
  src/_portaudiomodule.c(1367): error C2146: syntax error: missing ';' before identifier '_b'
  src/_portaudiomodule.c(1367): error C2065: '_b': undeclared identifier
  src/_portaudiomodule.c(1367): error C2065: '_a': undeclared identifier
  src/_portaudiomodule.c(1367): error C2059: syntax error: ')'
  src/_portaudiomodule.c(1371): error C2065: 'output_data': undeclared identifier
  src/_portaudiomodule.c(1371): warning C4022: 'memset': pointer mismatch for actual parameter 1
  src/_portaudiomodule.c(1376): error C2143: syntax error: missing ')' before 'string'
  src/_portaudiomodule.c(1376): error C2143: syntax error: missing '{' before 'string'
  src/_portaudiomodule.c(1376): error C2059: syntax error: 'string'
  src/_portaudiomodule.c(1378): error C2143: syntax error: missing '{' before ':'
  src/_portaudiomodule.c(1378): error C2059: syntax error: ':'
  src/_portaudiomodule.c(1385): error C2143: syntax error: missing ')' before '('
  src/_portaudiomodule.c(1385): error C2091: function returns function
  src/_portaudiomodule.c(1385): error C2059: syntax error: ')'
  src/_portaudiomodule.c(1385): error C2143: syntax error: missing ')' before ';'
  src/_portaudiomodule.c(1385): error C2373: '_Py_XDECREF': redefinition; different type modifiers
  f:\python\include\object.h(538): note: see declaration of '_Py_XDECREF'
  src/_portaudiomodule.c(1386): error C2143: syntax error: missing ')' before '('
  src/_portaudiomodule.c(1386): error C2091: function returns function
  src/_portaudiomodule.c(1386): error C2059: syntax error: ')'
  src/_portaudiomodule.c(1386): error C2143: syntax error: missing ')' before ';'
  src/_portaudiomodule.c(1386): error C2373: '_Py_XDECREF': redefinition; different type modifiers
  f:\python\include\object.h(538): note: see declaration of '_Py_XDECREF'
  src/_portaudiomodule.c(1387): error C2143: syntax error: missing ')' before '('
  src/_portaudiomodule.c(1387): error C2091: function returns function
  src/_portaudiomodule.c(1387): error C2059: syntax error: ')'
  src/_portaudiomodule.c(1387): error C2143: syntax error: missing ')' before ';'
  src/_portaudiomodule.c(1387): error C2373: '_Py_XDECREF': redefinition; different type modifiers
  f:\python\include\object.h(538): note: see declaration of '_Py_XDECREF'
  src/_portaudiomodule.c(1389): error C2371: 'PyGILState_Release': redefinition; different basic types
  f:\python\include\pystate.h(116): note: see declaration of 'PyGILState_Release'
  src/_portaudiomodule.c(1390): error C2059: syntax error: 'return'
  src/_portaudiomodule.c(1391): error C2059: syntax error: '}'
  error: command 'F:\\VisualStudio2017\\VC\\Tools\\MSVC\\14.12.25827\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
  ----------------------------------------
  ERROR: Failed building wheel for pyaudio
  Running setup.py clean for pyaudio
Failed to build pyaudio
Installing collected packages: send2trash, pyaudio, markdown, pyrsistent, jsonschema, pywin32, aqt
    Running setup.py install for pyaudio ... error
    ERROR: Command errored out with exit status 1:
     command: 'F:\anki\pyenv\Scripts\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Professional\\AppData\\Local\\Temp\\pip-install-_np3kykb\\pyaudio\\setup.py'"'"'; __file__='"'"'C:\\Users\\Professional\\AppData\\Local\\Temp\\pip-install-_np3kykb\\pyaudio\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Professional\AppData\Local\Temp\pip-record-kg2dohm3\install-record.txt' --single-version-externally-managed --compile --install-headers 'F:\anki\pyenv\include\site\python3.8\pyaudio'
         cwd: C:\Users\Professional\AppData\Local\Temp\pip-install-_np3kykb\pyaudio\
    Complete output (53 lines):
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.8
    copying src\pyaudio.py -> build\lib.win-amd64-3.8
    running build_ext
    building '_portaudio' extension
    creating build\temp.win-amd64-3.8
    creating build\temp.win-amd64-3.8\Release
    creating build\temp.win-amd64-3.8\Release\src
    F:\VisualStudio2017\VC\Tools\MSVC\14.12.25827\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DMS_WIN64=1 -IF:\anki\pyenv\include -IF:\Python\include -IF:\Python\include -IF:\VisualStudio2017\VC\Tools\MSVC\14.12.25827\ATLMFC\include -IF:\VisualStudio2017\VC\Tools\MSVC\14.12.25827\include "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" /Tcsrc/_portaudiomodule.c /Fobuild\temp.win-amd64-3.8\Release\src/_portaudiomodule.obj
    _portaudiomodule.c
    f:\python\include\pyconfig.h(117): warning C4005: 'MS_WIN64': macro redefinition
    f:\python\include\pyconfig.h(117): note: command-line arguments:  see previous definition of 'MS_WIN64'
    src/_portaudiomodule.c(44): warning C4005: 'min': macro redefinition
    C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt\stdlib.h(1290): note: see previous definition of 'min'
    src/_portaudiomodule.c(1367): error C2059: syntax error: '{'
    src/_portaudiomodule.c(1367): warning C4013: '__typeof__' undefined; assuming extern returning int
    src/_portaudiomodule.c(1367): error C2146: syntax error: missing ';' before identifier '_b'
    src/_portaudiomodule.c(1367): error C2065: '_b': undeclared identifier
    src/_portaudiomodule.c(1367): error C2065: '_a': undeclared identifier
    src/_portaudiomodule.c(1367): error C2059: syntax error: ')'
    src/_portaudiomodule.c(1371): error C2065: 'output_data': undeclared identifier
    src/_portaudiomodule.c(1371): warning C4022: 'memset': pointer mismatch for actual parameter 1
    src/_portaudiomodule.c(1376): error C2143: syntax error: missing ')' before 'string'
    src/_portaudiomodule.c(1376): error C2143: syntax error: missing '{' before 'string'
    src/_portaudiomodule.c(1376): error C2059: syntax error: 'string'
    src/_portaudiomodule.c(1378): error C2143: syntax error: missing '{' before ':'
    src/_portaudiomodule.c(1378): error C2059: syntax error: ':'
    src/_portaudiomodule.c(1385): error C2143: syntax error: missing ')' before '('
    src/_portaudiomodule.c(1385): error C2091: function returns function
    src/_portaudiomodule.c(1385): error C2059: syntax error: ')'
    src/_portaudiomodule.c(1385): error C2143: syntax error: missing ')' before ';'
    src/_portaudiomodule.c(1385): error C2373: '_Py_XDECREF': redefinition; different type modifiers
    f:\python\include\object.h(538): note: see declaration of '_Py_XDECREF'
    src/_portaudiomodule.c(1386): error C2143: syntax error: missing ')' before '('
    src/_portaudiomodule.c(1386): error C2091: function returns function
    src/_portaudiomodule.c(1386): error C2059: syntax error: ')'
    src/_portaudiomodule.c(1386): error C2143: syntax error: missing ')' before ';'
    src/_portaudiomodule.c(1386): error C2373: '_Py_XDECREF': redefinition; different type modifiers
    f:\python\include\object.h(538): note: see declaration of '_Py_XDECREF'
    src/_portaudiomodule.c(1387): error C2143: syntax error: missing ')' before '('
    src/_portaudiomodule.c(1387): error C2091: function returns function
    src/_portaudiomodule.c(1387): error C2059: syntax error: ')'
    src/_portaudiomodule.c(1387): error C2143: syntax error: missing ')' before ';'
    src/_portaudiomodule.c(1387): error C2373: '_Py_XDECREF': redefinition; different type modifiers
    f:\python\include\object.h(538): note: see declaration of '_Py_XDECREF'
    src/_portaudiomodule.c(1389): error C2371: 'PyGILState_Release': redefinition; different basic types
    f:\python\include\pystate.h(116): note: see declaration of 'PyGILState_Release'
    src/_portaudiomodule.c(1390): error C2059: syntax error: 'return'
    src/_portaudiomodule.c(1391): error C2059: syntax error: '}'
    error: command 'F:\\VisualStudio2017\\VC\\Tools\\MSVC\\14.12.25827\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'F:\anki\pyenv\Scripts\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Professional\\AppData\\Local\\Temp\\pip-install-_np3kykb\\pyaudio\\setup.py'"'"'; __file__='"'"'C:\\Users\\Professional\\AppData\\Local\\Temp\\pip-install-_np3kykb\\pyaudio\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Professional\AppData\Local\Temp\pip-record-kg2dohm3\install-record.txt' --single-version-externally-managed --compile --install-headers 'F:\anki\pyenv\include\site\python3.8\pyaudio' Check the logs for full command output.
make[1]: *** [Makefile:19: .build/run-deps] Error 1
make[1]: Leaving directory '/cygdrive/f/anki/qt'
make: *** [Makefile:47: develop] Error 2
make: Leaving directory '/cygdrive/f/anki'
```

This happens by these two problems on the `pyaudio` lib I fixed on the fork:
1. https://github.com/evandroforks/pyaudio/commit/e6ea652e81cb64c1debc48e1a248800ff5ca877c Fixed installation for Visual Studio 2017 on setup.py
1. https://github.com/evandroforks/pyaudio/commit/b37c601f488ef1ab3b85f0747ad5395c779fc87c Fixed min() macro not compilling with MSVC compiler

The alternative, would be:
1. To ask the user to install `pyaudio` on their system by following the instructions now on https://github.com/evandroforks/pyaudio
1. And then force the pyenv to inherit the installed `pyaudio` package from the installed python system

Right now I do not know how to do this. I will try to research it.

Reply:
After looking into it, I cannot find a way to not remove those lines. Even when using the pyaudio wheels from https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyaudio, it will not work because when you do `python -m venv pyenv`, you create a isolated environment from the system installation. Then, if I manually install the wheels from https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyaudio in my system, before calling `python -m venv pyenv`, it will not work because the `pyenv` is going to ignore the packages I have installed system wide and it is going to download the `pyaudio` from pypi.

After researching a few links:
1. https://stackoverflow.com/questions/55600132/installing-local-packages-with-python-virtualenv-system-site-packages
1. https://github.com/pypa/pip/issues/1408 pip and virtualenv --system-site-packages
1. https://stackoverflow.com/questions/14571454/virtualenv-specifing-which-packages-to-use-system-wide-vs-local
1. https://github.com/pyenv/pyenv/issues/452 pyenv not able to pickup packages from system if not present in local
1. https://stackoverflow.com/questions/12079607/make-virtualenv-inherit-specific-packages-from-your-global-site-packages
1. https://stackoverflow.com/questions/52921165/make-a-pycharm-project-inheriting-global-site-packages-after-creating-the-proje

The most reliable option to remove the line `INSTALL_PYAUDIO := python -m pip install git+https...` on is to run this command right after calling `python -m venv pyenv`:
```make
sed -iv -- "s/include-system-site-packages\s*=\s*false/include-system-site-packages = true/g" pyenv/pyvenv.cfg
```

This is the only way to make `pyenv` see the system wide packages, without causing the packages installed inside `pyenv` to be installed system wide (as would happen if we called `python -m venv pyenv` with `--system-site-packages`.

But now my installation is breaking because it is not isolated from my system python anymore, and it is complaining some anki dependecy requires `mypy 0.560 has requirement typed-ast<1.2.0,>=1.1.0, but you'll have typed-ast 1.3.4 which is incompatible.` and when it try to install `typed-ast 1.1.2` it fails down with error: `Cannot open include file: 'pgenheaders.h'` which was only fixed on (https://github.com/python/typed_ast/issues/97). This error with `typed-ast` happens because on Python 3.8 or superior have removed the `pgenheaders.h` and only `typed-ast` version `1.3.5` or superior has a fix for it.

Reply:
I managed to fix the error of the `typed-ast` by running `pip install mypy -U`.

Now the error is:
```
make[1]: Entering directory '/cygdrive/f/anki/qt'
./tools/build_ui.sh
Generating forms..
 * aqt/forms/about.py
Traceback (most recent call last):
  File "f:\python\lib\runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "f:\python\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "F:\Python\Scripts\pyuic5.exe\__main__.py", line 4, in <module>
  File "f:\python\lib\site-packages\PyQt5\uic\pyuic.py", line 26, in <module>
    from PyQt5 import QtCore
ModuleNotFoundError: No module named 'PyQt5.sip'
make[1]: *** [Makefile:27: .build/ui] Error 1
make[1]: Leaving directory '/cygdrive/f/anki/qt'
make: *** [Makefile:50: develop] Error 2
make: Leaving directory '/cygdrive/f/anki'
```

I already tried following the instruction on https://github.com/Electron-Cash/Electron-Cash/issues/892, now I got the error:
```
set -eo pipefail && \
. "pyenv/Scripts/activate" && \
echo "Starting Anki..."; \
python qt/runanki
Starting Anki...
Traceback (most recent call last):
  File "qt/runanki", line 3, in <module>
    import aqt
  File "F:\anki\qt\aqt\__init__.py", line 21, in <module>
    from aqt.qt import *
  File "F:\anki\qt\aqt\qt.py", line 16, in <module>
    from PyQt5.QtWebEngineWidgets import *  # type: ignore
ImportError: DLL load failed while importing QtWebEngineWidgets: The specified procedure could not be found.
make: *** [Makefile:58: run] Error 1
make: Leaving directory '/cygdrive/f/anki'
```

Reply:
After finding https://stackoverflow.com/questions/37876987/cannot-import-qtwebkitwidgets-in-pyqt5 I managed to fix the last error `ImportError: DLL load failed while importing QtWebEngineWidgets` by running the command `pip install PyQtWebEngine`, now I am finally running anki with the latest changes you requested.

I would rather do not have to use the system packages (with `sed -iv -- "s/include-system-site-packages\s*=\s*false/include-system-site-packages = true/g" pyenv/pyvenv.cfg` on the Makefile) and have the `pyenv` environment, completely isolated from my system python packages installation, as it was before you request to remove the `INSTALL_PYAUDIO := python -m pip install git+https...` and use the python wheels from https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyaudio.

Reply:
Due all this problems of packages versions incompatibilities, I reverted back into using a isolated pyenv on Windows. 

Now I update the installation process and the user can now choose between:
1. Do not do anything, i.e., do no define the `ANKI_EXTRA_PIP` environment variable and nothing special will be done. Which will probably cause the error of missing `portaudio.lib` file.
1. The user can define the `ANKI_EXTRA_PIP` as `set "ANKI_EXTRA_PIP=python -m pip install full/path/to/PyAudio‑0.2.11‑cp38‑cp38‑win_amd64.whl"`, i.e., he can download a wheel from https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyaudio and let anki build process install it.
1. The user can follow my instructions to build his own `portaudio.lib` on https://github.com/evandroforks/pyaudio and define the `ANKI_EXTRA_PIP` variable as `set "ANKI_EXTRA_PIP=python -m pip install git+https://github.com/evandroforks/pyaudio"` or some place else, before before running anki.

Note, that it is only required to define the optional environment variable `ANKI_EXTRA_PIP` on the first time the `pyenv` is initialized. Now the user can install whatever `pyaudio` lib he want to. This trick even works for any library on Linux too (except that the syntax is `export var="..."` instead of `set "var=..."`.